examples/nlp/nl_image_search.py

```diff
@@ -0,0 +1,605 @@
+    embeddings = bert(bert_inputs)["pooled_output"]
+    # Project the embeddings produced by the model.
+    outputs = project_embeddings(
+        embeddings, num_projection_layers, projection_dims, dropout_rate
+    )
+    # Create the text encoder model.
+    encoder = keras.Model(inputs, outputs, name="text_encoder")
+    return encoder
+
+
+"""
+## Implement contrastive similarity loss
+
+To calculate the loss, we compute the pairwise dot-product similarity between
+each `caption_i` and `images_j` in the batch as the predictions.
+The target similarity between `caption_i`  and `image_j` is computed as
+the average of the (dot-product similarity between `caption_i` and `caption_j`)
+and (the dot-product similarity between `image_i` and `image_j`).
+Then we use crossentropy to compute the loss between the targets and the predictions.
+"""
+
+
+class ContrastiveSimilarityLoss(keras.losses.Loss):
+    def __init__(self, temperature, **kwargs):
+        super(ContrastiveSimilarity, self).__init__(**kwargs)
+        self.temperature = temperature
+
+    def __call__(self, y_true, y_pred, sample_weight=None):
+        # Delete y_true as it is not used.
+        del y_true
+        # Extract the caption_embeddings and image_embeddings from the model output.
+        caption_embeddings, image_embeddings = tf.unstack(y_pred)
+        # logits[i][j] is the dot_similarity(caption_i, image_j).
+        logits = (
+            tf.matmul(caption_embeddings, image_embeddings, transpose_b=True)
+            / self.temperature
+        )
+        # images_similarity[i][j] is the dot_similarity(image_i, image_j).
+        images_similarity = tf.matmul(
+            image_embeddings, image_embeddings, transpose_b=True
+        )
+        # captions_similarity[i][j] is the dot_similarity(caption_i, caption_j).
+        captions_similarity = tf.matmul(
+            caption_embeddings, caption_embeddings, transpose_b=True
+        )
+        # targets[i][j] = avarage dot_similarity(caption_i, caption_j) and dot_similarity(image_i, image_j).
+        targets = keras.activations.softmax(
+            (captions_similarity + images_similarity) / (2 * self.temperature)
+        )
+        # Compute the loss for the captions using crossentropy
+        captions_loss = keras.losses.categorical_crossentropy(
+            y_true=targets, y_pred=logits, from_logits=True
+        )
+        # Compute the loss for the images using crossentropy
+        images_loss = keras.losses.categorical_crossentropy(
+            y_true=tf.transpose(targets), y_pred=tf.transpose(logits), from_logits=True
+        )
+        # Compute the average between the captions loss and the images loss
+        loss = (captions_loss + images_loss) / 2
+        # Return the mean of the loss over the batch.
+        return tf.math.reduce_mean(loss)
+
+
+"""
+## Implement the dual encoder
+"""
+
+
+def create_dual_encoder(text_encoder, vision_encoder):
```

Comment:
This could be a model subclass taking the 2 submodels in its constructor and featuring a custom `train_step`. The model has basically no topology so it would make a lot of sense. Its `call` pass would just be:

```
def call(self, captions, images):
    # Get the embeddings for the captions.
    caption_embeddings = self.text_encoder(captions)
    # Get the embeddings for the images.
    image_embeddings = self.vision_encoder(images)
    # Stack the caption_embeddings and image_embeddings to produce
    # the output
    return tf.stack([caption_embeddings, image_embeddings])
```

The `train_step` would be the content of the contrastive loss. You may have to implement an `evaluate_step` and `predict_step` as well.

Reply:
I might recommend `def call(self, inputs)` then unpacking `captions`/`images` from the inputs, which enables `predict` to be used within a custom `predict_step`.

Reply:
Thank you for the suggestion. I tried the custom model/training loop approach, but it doesn't be really needed, nor does it simplify the code. 

Note that the dual encoder is discarded once  the two encoders are trained, so we don't need the `predict_step` nor `evaluate_step`. The reasons is that, in inference time, we don't provide a query and an image to we predict their similarity (in which case, the `predict_step` would make sense), rather we provide a query and we search for the most similar image. So forward pass of the training process is irrelevant in prediction.

What do you think?

Reply:
I think we should do this: https://gist.github.com/fchollet/196b4b48ce11569aa5e256dc6f274f6b (caveat: I haven't tested the code)

It is more extensible, and it does not involve any "hacks" (like fake labels), so it will look nicer to people who want maximum flexibility. It also takes fewer lines of code.